adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpConnectionLimitStrategy.java

```diff
@@ -0,0 +1,83 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.adapter.amqp;
+
+import org.eclipse.hono.service.limiting.MemoryBasedConnectionLimitStrategy;
+
+/**
+ * This class calculates the recommended limit of concurrent connections, based on the runtime's memory and the AMQP
+ * session window size ({@link org.eclipse.hono.adapter.amqp.AmqpAdapterProperties#getMaxSessionWindowSize()}).
+ *
+ */
+public class AmqpConnectionLimitStrategy extends MemoryBasedConnectionLimitStrategy {
+
+    /**
+     * The minimum amount of memory that the adapter requires to run.
+     */
+    private static final int MINIMAL_MEMORY = 100_000_000; // 100MB: minimal memory necessary for startup
+
+    /**
+     * The amount of memory required for each connection.
+     */
+    private static final int MEMORY_PER_CONNECTION = 20_000; // 20KB: expected avg. memory consumption per connection
+
+    private final int maxSessionWindowSize;
+
+    /**
+     * Creates an instance that calculates the recommended limit dependent on the runtime's memory and the maximum AMQP
+     * session window size. The session window size of the given configuration is used if provided, otherwise the
+     * default session window size is used.
+     * 
+     * @param config The configuration of the protocol adapter or {@code null}.
+     */
+    public AmqpConnectionLimitStrategy(final AmqpAdapterProperties config) {
+
+        if (config != null) {
+            maxSessionWindowSize = config.getMaxSessionWindowSize();
+        } else {
+            maxSessionWindowSize = AmqpAdapterProperties.DEFAULT_MAX_FRAME_SIZE_BYTES
+                    * AmqpAdapterProperties.DEFAULT_MAX_SESSION_FRAMES;
+        }
+    }
+
+    /**
+     * Use only for tests!
+     * 
+     * @param maxMemory The amount of memory to test against.
+     * @param maxSessionWindowSize The maximum AMQP session window size to test against.
+     */
+    AmqpConnectionLimitStrategy(final long maxMemory, final int maxSessionWindowSize) {
+        this.maxMemory = maxMemory;
+        this.maxSessionWindowSize = maxSessionWindowSize;
+    }
+
+    @Override
+    protected long getMinimalMemory() {
+        return MINIMAL_MEMORY;
+    }
+
+    /**
+     * Returns a recommended limit of concurrent connections for the given maximum amount of memory. An amount of 100MB
+     * should be reserved for the startup of the adapter. Per connection the maximum AMQP session window size + 20KB are
+     * needed.
+     * <p>
+     *
+     * @param maxMemory The maximum amount of memory for the JVM in bytes as returned by {@link Runtime#maxMemory()}.
+     * @return The recommended maximum connection limit.
+     */
+    @Override
+    protected long calculateRecommendedLimit(final long maxMemory) {
+        return (maxMemory - MINIMAL_MEMORY) / (MEMORY_PER_CONNECTION + maxSessionWindowSize);
```

Comment:
isn't this formula always the same for all connection oriented protocols?
I wonder if `MemoryBasedConnectionLimitStrategy` couldn't implement this method in a generic way:
```
return (maxMemory - MINIMAL_MEMORY) / (getMemoryRequiredPerConnection());
```
and let subclasses only implement *getMemoryRequiredPerConnection* ...

Reply:
You are right. Then I would go one step further: there is no need for`MemoryBasedConnectionLimitStrategy` to be extended, since the calculation is done there based on static values (for minimal memory and memory required per connection), which could be constructor parameters.

Reply:
sounds even better :+1: